Framework/Algorithms/src/CarpenterSampleCorrection.cpp

```diff
@@ -0,0 +1,175 @@
+#include "MantidAPI/WorkspaceUnitValidator.h"
+#include "MantidDataObjects/EventWorkspace.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidGeometry/IComponent.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/Material.h"
+#include "MantidKernel/PhysicalConstants.h"
+
+#include <stdexcept>
+
+namespace Mantid {
+namespace Algorithms {
+DECLARE_ALGORITHM(CarpenterSampleCorrection) // Register the class
+                                             // into the algorithm
+                                             // factory
+
+using namespace Kernel;
+using namespace API;
+using Mantid::DataObjects::EventList;
+using Mantid::DataObjects::EventWorkspace;
+using Mantid::DataObjects::EventWorkspace_sptr;
+using Mantid::DataObjects::WeightedEventNoTime;
+using Mantid::HistogramData::HistogramX;
+using Mantid::HistogramData::HistogramY;
+using Mantid::HistogramData::HistogramE;
+using std::vector;
+using namespace Mantid::PhysicalConstants;
+using namespace Geometry;
+
+// Constants required internally only, so make them static. These are
+// Chebyshev expansion coefficients copied directly from Carpenter 1969 Table 1
+// clang-format off
+const std::string CarpenterSampleCorrection::name() const {
+  return "CarpenterSampleCorrection";
+}
+
+int CarpenterSampleCorrection::version() const { return 1; }
+
+const std::string CarpenterSampleCorrection::category() const {
+  return "CorrectionFunctions\\AbsorptionCorrections";
+}
+
+/**
+ * Initialize the properties to default values
+ */
+void CarpenterSampleCorrection::init() {
+  // The input workspace must have an instrument and units of wavelength
+  auto wsValidator = boost::make_shared<CompositeValidator>();
+  wsValidator->add<WorkspaceUnitValidator>("Wavelength");
+  wsValidator->add<InstrumentValidator>();
+
+  declareProperty(make_unique<WorkspaceProperty<API::MatrixWorkspace> >(
+                      "InputWorkspace", "", Direction::Input, wsValidator),
+                  "The name of the input workspace.");
+  declareProperty(make_unique<WorkspaceProperty<API::MatrixWorkspace> >(
+                      "OutputWorkspace", "", Direction::Output),
+                  "The name of the output workspace.");
+
+  declareProperty("AttenuationXSection", 2.8, "Coefficient 1, absorption cross "
+                                              "section / 1.81 if not set with "
+                                              "SetSampleMaterial");
+  declareProperty("ScatteringXSection", 5.1, "Coefficient 3, total scattering "
+                                             "cross section if not set with "
+                                             "SetSampleMaterial");
+  declareProperty("SampleNumberDensity", 0.0721,
+                  "Coefficient 2, density if not set with SetSampleMaterial");
+  declareProperty("CylinderSampleRadius", 0.3175, "Sample radius, in cm");
+}
+
+/**
+ * Execute the algorithm
+ */
+void CarpenterSampleCorrection::exec() {
```

Comment:
Should set up progress bar updating. This can be done when you create the child algorithm specify two extra parameters `createChildAlgorithm("Divide", startPercent, stopPercent)` then the child will update things appropriately.